File: correct-crypto/ruby/hmac_hash.rb
# frozen_string_literal: true
# Correct SHA-256 hashing and HMAC-SHA-256 with constant-time verification (OpenSSL).
require 'openssl'
require 'digest'

def sha256_hex(data)
  Digest::SHA256.hexdigest(data)
end

def hmac_sha256(key, data)
  OpenSSL::HMAC.digest('SHA256', key, data)
end

def verify(key, data, tag)
  OpenSSL.fixed_length_secure_compare(hmac_sha256(key, data), tag)
end

File: correct-crypto/python/hmac_hash.py
#!/usr/bin/env python3
"""Correct hashing (SHA-256) and message authentication (HMAC-SHA-256, stdlib)."""
import hashlib
import hmac


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def hmac_sha256(key: bytes, data: bytes) -> bytes:
    return hmac.new(key, data, hashlib.sha256).digest()


def verify(key: bytes, data: bytes, tag: bytes) -> bool:
    return hmac.compare_digest(hmac_sha256(key, data), tag)   # constant-time


if __name__ == "__main__":
    key = b"secret-mac-key"
    tag = hmac_sha256(key, b"authenticate me")
    assert verify(key, b"authenticate me", tag)
    print("SHA-256:", sha256_hex(b"hello"))
    print("HMAC-SHA-256 verify OK")
